esp-hal/src/dma/buffers.rs

```diff
@@ -1562,3 +1564,364 @@ impl DerefMut for DmaLoopBuf {
         self.buffer
     }
 }
+
+/// A Preparation that masks itself as a DMA buffer.
+///
+/// Fow low level use, where none of the pre-made buffers really fit.
+///
+/// This type likely never should be visible outside of esp-hal.
+pub(crate) struct NoBuffer(Preparation);
+impl NoBuffer {
+    fn prep(&self) -> Preparation {
+        Preparation {
+            start: self.0.start,
+            direction: self.0.direction,
+            #[cfg(psram_dma)]
+            accesses_psram: self.0.accesses_psram,
+            burst_transfer: self.0.burst_transfer,
+            check_owner: self.0.check_owner,
+            auto_write_back: self.0.auto_write_back,
+        }
+    }
+}
+unsafe impl DmaTxBuffer for NoBuffer {
+    type View = Self;
+
+    fn prepare(&mut self) -> Preparation {
+        self.prep()
+    }
+
+    fn into_view(self) -> Self::View {
+        self
+    }
+
+    fn from_view(view: Self::View) -> Self {
+        view
+    }
+}
+unsafe impl DmaRxBuffer for NoBuffer {
+    type View = Self;
+
+    fn prepare(&mut self) -> Preparation {
+        self.prep()
+    }
+
+    fn into_view(self) -> Self::View {
+        self
+    }
+
+    fn from_view(view: Self::View) -> Self {
+        view
+    }
+}
+
+/// Prepares data unsafely to be transmitted via DMA.
+///
+/// `block_size` is the requirement imposed by the peripheral that receives the data. It
+/// ensures that the DMA will not try to copy a partial block, which would cause the RX DMA (that
+/// moves results back into RAM) to never complete.
+///
+/// The function returns the DMA buffer, and the number of bytes that will be transferred.
+///
+/// # Safety
+///
+/// The caller must keep all its descriptors and the buffers they
+/// point to valid while the buffer is being transferred.
+#[cfg_attr(not(aes_dma), expect(unused))]
+pub(crate) unsafe fn prepare_for_tx(
+    descriptors: &mut [DmaDescriptor],
+    mut data: NonNull<[u8]>,
+    block_size: usize,
+) -> (NoBuffer, usize) {
+    let chunk_size = BurstConfig::DEFAULT
+        .max_chunk_size_for(unsafe { data.as_ref() }, TransferDirection::Out)
+        .min(4096 - block_size); // Whichever is stricter, data location or peripheral requirements
```

Comment:
I had a comment about this somewhere but this `min(...)` isn't needed. It has no influence on the amount or contents of the data the TX channel transmits.
All it does is increase the number of required descriptors.

Reply:
Yes I understand that the *data* needs to be a multiple of the block size, what I don't understand is why you need to use more descriptors to do the exact same thing.

Simply using the correct data length (as suppose to `chunk_size`) would achieve this goal no?

Reply:
I suppose, but I suspect "simply" calculating the right data length is more complicated than this is. And this code is complicated enough as it is, even if we don't take the other side's burst config into account yet. I'm not entirely worried about the maximum 1.5% loss of efficiency in terms of DMA descriptor use.

Reply:
Yeah, I initially wasn't worried about the descriptor efficiency either but then I remembered people often use code as documentation (given how quirky some of these chips can be), and I don't want this snippet to be misunderstood by future readers.

To save you the mental workload.

```Rust
let max_transfer_size = chunk_size * descriptors.len();
let max_aligned_transfer_size = (max_transfer_size / block_size) * block_size; // usize::prev_multiple_of
let data_len = data.len().min(max_aligned_transfer_size);
```

Side note: The use of `min` in the existing code is a bit sus but I figure `data.len()` is already aligned when this function is called.